dotnet sln: Add error message for misplaced solution argument

ZeroErrors opened the pull request:
Fixes https://github.com/dotnet/sdk/issues/17686
(Created a new issue for one of the causes not solved in this PR https://github.com/dotnet/sdk/issues/18607)

## Change log
- Added basic check to provide more descriptive message when solution argument is misplaced in `dotnet sln add` and `dotnet sln remove`.
- Added basic command suggestion output to help guide the user.

## Testing
Unit tests included.

To test manually setup a solution and a project something like
```
dotnet new sln
dotnet new console --output src/todo-app
dotnet sln add src/todo-app --solution-folder todo
```

Then running `dotnet sln add <project>.sln src/todo-app` will output
```
Solution argument 'App.sln' is misplaced.
Did you mean?
  dotnet sln <project>.sln add src/todo-app
```
and then the help text

ZeroErrors committed:
dotnet sln: Add error message for misplaced solution argument

sfoslund commented on src/Cli/dotnet/commands/dotnet-sln/remove/Program.cs:
```diff
@@ -18,16 +18,27 @@ internal class RemoveProjectFromSolutionCommand : CommandBase
+                throw new GracefulException(new string[]
+                {
+                    string.Format(CommonLocalizableStrings.SolutionArgumentMisplaced, slnFile),
+                    CommonLocalizableStrings.DidYouMean,
```
@KathleenDollard what do you think of the user experience here? The 'did you mean...' message feels odd to me, I think if anything we should provide an error message educating the user on the difference between a sln and proj as well as print the help here.

ZeroErrors replied:
I personally just find it faster to resolve problems when a tool can figure out and tell me what it expected instead of needing to spend time to figure out what it wanted but I can just remove the "Did you mean?" part if that's not what is wanted.
If you think its worth talking about it can be split out into an issue and discussed there instead maybe?

The help usage gets printed on stdout as part of `GracefulException` getting thrown.

sfoslund commented on src/Tests/dotnet-sln.Tests/GivenDotnetSlnRemove.cs:
```diff
@@ -735,5 +736,26 @@ public void WhenProjectIsRemovedThenDependenciesOnProjectAreAlsoRemoved()
+        public void WhenSolutionIsPassedAsProjectItPrintsSuggestionAndUsage()
+        {
+            var projectDirectory = _testAssetsManager
+                .CopyTestAsset("TestAppWithSlnAndCsprojFiles", identifier: "WhenSolutionIsPassedAsProject")
```
I don't think the identifier is needed here (or any of the other tests)

ZeroErrors replied:
Ok will do, wasn't sure. I couldn't find any documentation on how this is supposed to work so have just been guessing.

nagilson replied:
The identifier is a parameter used for uniquely identifying [Theory] tests vs [Fact] tests because they get put into a flat folder with the test name into a machine that runs them in parallel. Thanks for resolving it. I added documentation for this stuff a few months back as I had the same confusion!

sfoslund commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -43,27 +49,39 @@ internal class AddProjectToSolutionCommand : CommandBase
                 _relativeRootSolutionFolders = null;
             }
+
+            var slnFile = _arguments.FirstOrDefault(path => path.EndsWith(".sln"));
```
We can factor this and above arguments parsing and checking count into a private method called something like "ParseAndValidateArguments"

sfoslund commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -43,27 +49,39 @@ internal class AddProjectToSolutionCommand : CommandBase
+            if (slnFile != null)
+            {
+                string args;
+                if (_inRoot) args = $"--{SlnAddParser.InRootOption.Name} ";
```
nit: avoid one line if statements to keep code style consistency in the repo (Although I think we'll probably want to remove this anyways and instead give an error that doesn't include these args)

sfoslund commented on src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -43,27 +49,39 @@ internal class AddProjectToSolutionCommand : CommandBase
                     MsbuildProject.GetProjectFileFromDirectory(fullPath).FullName :
                     fullPath;
-            }).ToList();
+            }).ToArray();
```
Why is this change necessary? I think it should be reverted.

ZeroErrors replied:
It was part of changes I was working on trying to deal with https://github.com/dotnet/sdk/issues/18607 but it became a deeper problem than I wanted to address in this PR which is why I made a new issue.

I'll revert it and the same change in `dotnet sln remove` since its not needed for this PR.

sfoslund commented on src/Cli/dotnet/commands/dotnet-sln/remove/Program.cs:
```diff
@@ -18,16 +18,27 @@ internal class RemoveProjectFromSolutionCommand : CommandBase
 
-            _fileOrDirectory = parseResult.ValueForArgument<string>(SlnCommandParser.SlnArgument);
+            var slnFile = _arguments.FirstOrDefault(path => path.EndsWith(".sln"));
+            if (slnFile != null)
```
Same as above, we can factor this section into a method. In fact, we can probably reuse the same method across both commands.

sfoslund commented on src/Cli/dotnet/commands/dotnet-sln/remove/Program.cs:
```diff
@@ -43,7 +54,7 @@ public override int Execute()
                         fullPath
                 );
-            });
+            }).ToArray();
```
nit: unnecessary change, revert

sfoslund reviewed (commented):
Thanks for your contribution! Overall, I think it makes sense for us to give a better error for this scenario but I'd like @KathleenDollard to give her input on the specific wording of the error. Otherwise the code quality looks good to me, thanks for adding tests!

ZeroErrors committed:
dotnet sln: Remove unneeded identifier from unit tests

ZeroErrors committed:
dotnet sln: Remove unnecessary changes

marcpopMSFT committed:
Merge remote-tracking branch 'origin/main' into dotnet-sln-arg-misplaced-error

marcpopMSFT committed:
Refactor the validation code into it's own shared method per PR feedback
Update the string text to be more consistent with Templating per PM
Fix up the tests for changes in main

marcpopMSFT committed:
Update one more test that I missed

marcpopMSFT commented:
Catching up on old PRs. I merged with main, made the refactoring that was suggested in the last PR review, fixed the tests, and reviewed the text with PM and aligned it with templating engine text in a similar "suggested command" scenario: https://github.com/dotnet/sdk/blob/main/src/Cli/Microsoft.TemplateEngine.Cli/LocalizableStrings.resx#L799

nagilson commented on src/Tests/dotnet-sln.Tests/GivenDotnetSlnAdd.cs:
```diff
@@ -1225,5 +1225,68 @@ public void WhenSolutionFolderIsPassedWithDirectorySeparatorFolderStructureIsCor
         }
+
+        [Fact]
+        public void WhenSolutionIsPassedAsProjectItPrintsSuggestionAndUsage()
```
This is somewhat of a nit.. but all 15 lines of these three (really four) tests are identical minus the arguments passed in to them. I think we should instead make a helper method that just takes in the execute arguments and expected error message? Or you could turn this into a theory. I prefer the helper method because it lets you keep the better test names.

nagilson commented on src/Cli/dotnet/commands/dotnet-sln/SlnArgumentValidator.cs:
```diff
@@ -0,0 +1,59 @@
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.DotNet.Cli;
```
I appreciate that you refactored this into its own class. Thank you.

nagilson reviewed (commented):
This is a really nice to have UX improvement. Thank you for your contribution once again. :) Even if we are pretty late to getting to it! And thanks Marc for giving it a refresh.

marcpopMSFT committed:
Collapse test code together per PR feedback

marcpopMSFT commented:
Thank you for your contribution!
